sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,600 @@
+   following while the tx fee is higher than the total cost incurred
+   by walking to the next node in the AST:
+   a. If the next node calls a contract method, then verify that
+      the contract exists and the method arguments match the contract's
+      `define-public` signature. If not, then `RETURN INVALID`.
+   b. Compute the runtime cost of each node in the AST, adding it
+      to the function's cost analysis.
+7. Find all `define-map` calls to find all tables that need to
+   exist. Each step in this incurs a small fee.
+8. Create all the tables if the cost of creating them is smaller than
+   the remaining tx fee. If not, then RETURN INVALID.
+9. `RETURN OK`
+
+## Database Requirements and Transaction Accounting
+
+The smart contract VM needs to interact with a database somewhat
+directly: the effects of an `insert-entry!` or `set-entry!` call are
+realized later in the execution of the same transaction. The database
+will need to support fairly fine-grained rollbacks as some contract
+calls within a transaction's execution may fail, triggering a
+rollback, while the transaction execution continues and successfully
+completes other database operations.
+
+The database API provided to the smart contract VM, therefore, must be
+capable of (1) quickly responding to `fetch-entry` queries, which are
+essentially simply key-value _gets_ on the materialized view of the
+operation log. The operation log itself is simply a log of the
+`insert-entry!` and `set-entry!` calls. In addition to these
+operations, the smart contract VM will be making token transfer calls.
+The databasse log should track those operations as well.
+
+In order to aid in accounting for the database operations created by a
+given transaction, the underlying database should store, with each
+operation entry, the corresponding transaction identifier. This will
+be expanded in a future SIP to require the database to store enough
+information to reconstruct each block, such that the blocks can be
+relayed to bootstrapping peers.
+
+# Measuring Transaction Costs for Fee Collection
+
+Our smart contracting language admits static analysis to determine
+many properties of transactions _before_ executing those
+transactions. In particular, it allows for the VM to count the total
+number of runtime operations required, the maximum amount of database
+writes, and the maximum number of calls to any expensive primitive
+functions like database reads or hash computations. Translating that
+information into transaction costs, however, requires more than simply
+counting those operations. It requires translating the operations into
+a single cost metric (something like gas in Ethereum). Then, clients
+can set the fee rate for that metric, and pay the corresponding
+transaction fee. Notably, unlike Turing-complete smart contracting
+languages, any such fees are known _before_ executing the transaction,
+such that clients will no longer need to estimate gas fees.
```

Comment:
But they will need to estimate Bitcoin-like fees for bribing miners to include the transaction in the chain (much easier than estimating gas, but still something that has to be done at runtime).

Reply:
yep, added